File: crates/gglib-cli/src/daemon_client/remote.rs
//! The remote tunnel's calls on a [`DaemonHandle`] (ADR 0012).

use std::time::Duration;

use anyhow::Result;

use super::wire::{
    RemoteConnectBody, RemoteConnectDto, RemoteDeviceDto, RemoteEnableBody, RemoteEnableDto,
    RemoteForgottenDto, RemoteStatusDto,
};
use super::{DaemonHandle, paths};

impl DaemonHandle {
    /// Bring the tunnel up and arm a pairing. The response is the only time
    /// the ticket and the code are ever handed out.
    ///
    /// Long timeout: the daemon waits for the endpoint to find a relay before
    /// minting the ticket, and may wait one settings-cache window for a
    /// freshly minted key to take effect on the local proxy.
    pub(crate) async fn remote_enable(&self, body: &RemoteEnableBody) -> Result<RemoteEnableDto> {
        let response = self
            .post(paths::REMOTE_ENABLE_PATH)
            .json(body)
            .timeout(Duration::from_secs(45))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// Take the tunnel down (idempotent on the daemon side).
    pub(crate) async fn remote_disable(&self) -> Result<RemoteStatusDto> {
        let response = self
            .post(paths::REMOTE_DISABLE_PATH)
            .json(&serde_json::json!({}))
            .timeout(Duration::from_secs(15))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// Mint a key for one new device and offer a code that hands it over.
    ///
    /// The same response `enable` gives, because it is the same answer: the
    /// service hands back the whole session, and the ticket in it is half of
    /// what the pairing screen draws.
    ///
    /// Shorter timeout than `remote_enable`'s. Nothing here waits on a relay
    /// — the tunnel is already up or this is refused — but two stores are
    /// written and the edge is told, so it is not instant either.
    pub(crate) async fn remote_invite(&self) -> Result<RemoteEnableDto> {
        let response = self
            .post(paths::REMOTE_INVITE_PATH)
            .json(&serde_json::json!({}))
            .timeout(Duration::from_secs(20))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// Every device this machine has issued a key to.
    pub(crate) async fn remote_devices(&self) -> Result<Vec<RemoteDeviceDto>> {
        let response = self
            .get(paths::REMOTE_DEVICES_PATH)
            .timeout(Duration::from_secs(5))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// Stop admitting one device. `forgotten` is false when this machine
    /// held nothing under that name, which is an answer rather than an error.
    pub(crate) async fn remote_forget(&self, device: &str) -> Result<RemoteForgottenDto> {
        let response = self
            .delete(&paths::remote_forget_path(device))
            .timeout(Duration::from_secs(15))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// The tunnel's status.
    pub(crate) async fn remote_status(&self) -> Result<RemoteStatusDto> {
        let response = self
            .get(paths::REMOTE_STATUS_PATH)
            .timeout(Duration::from_secs(5))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// Reach another machine: bind a loopback port here that is its proxy.
    ///
    /// Long timeout: dialling may wait for a hole punch, and a first pairing
    /// makes one more request through the tunnel before answering.
    pub(crate) async fn remote_connect(
        &self,
        body: &RemoteConnectBody,
    ) -> Result<RemoteConnectDto> {
        let response = self
            .post(paths::REMOTE_CONNECT_PATH)
            .json(body)
            .timeout(Duration::from_secs(60))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// Close the loopback port (idempotent on the daemon side).
    pub(crate) async fn remote_disconnect(&self) -> Result<RemoteStatusDto> {
        let response = self
            .post(paths::REMOTE_DISCONNECT_PATH)
            .json(&serde_json::json!({}))
            .timeout(Duration::from_secs(15))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }

    /// Stop the far daemon through the tunnel, then disconnect. The
    /// confirmation word is the daemon route's contract, not this client's
    /// idea: the CLI has already asked the person.
    pub(crate) async fn remote_kill(&self) -> Result<RemoteStatusDto> {
        let response = self
            .post(paths::REMOTE_KILL_PATH)
            .json(&serde_json::json!({ "confirm": "shutdown" }))
            .timeout(Duration::from_secs(30))
            .send()
            .await?;
        Ok(Self::expect_ok(response).await?.json().await?)
    }
}

File: src/services/transport/api/remote.ts
/**
 * Remote tunnel API module (ADR 0012).
 *
 * Eight calls, three groups. `enable`/`disable`/`getRemoteStatus` are this
 * machine as the desktop: the tunnel in front of its own proxy.
 * `connect`/`disconnect`/`kill` are this machine as the laptop: a loopback
 * port here that is another machine's proxy. `invite`/`forgetDevice` are who
 * may use the desktop's tunnel — a key per device, so retiring one leaves the
 * others connected. Every one goes to the daemon over HTTP — there are no
 * Tauri commands for the tunnel, so web and desktop take the same path.
 *
 * There is deliberately no `listDevices` here, though the daemon answers
 * `GET /api/remote/devices`: the roster rides `getRemoteStatus`, which is
 * re-read when the panel opens and on every tunnel event, and a second way
 * to read it is a second answer that can disagree with the `enabled` beside
 * it. After a `forgetDevice`, await a status refresh.
 */

import { del, get, post } from './client';
import {
  REMOTE_CONNECT_PATH,
  REMOTE_DEVICES_PATH,
  REMOTE_DISABLE_PATH,
  REMOTE_DISCONNECT_PATH,
  REMOTE_ENABLE_PATH,
  REMOTE_INVITE_PATH,
  REMOTE_KILL_PATH,
  REMOTE_STATUS_PATH,
} from '../../api/routes';
import type {
  RemoteConnectBody,
  RemoteConnectResponse,
  RemoteEnableBody,
  RemoteEnableResponse,
  RemoteForgotten,
  RemoteStatus,
} from '../types/remote';

/** The tunnel as the daemon reports it: both sides, fingerprints only. */
export async function getRemoteStatus(): Promise<RemoteStatus> {
  return get<RemoteStatus>(REMOTE_STATUS_PATH);
}

/**
 * Bring the tunnel up in front of this machine's proxy and arm a pairing.
 *
 * The response is the only time the ticket and the pairing code are handed
 * out; a second call while enabled is a 409, not a re-read. Show it once,
 * and let it go when a device pairs or the code expires.
 */
export async function enableRemote(body: Partial<RemoteEnableBody> = {}): Promise<RemoteEnableResponse> {
  return post<RemoteEnableResponse>(REMOTE_ENABLE_PATH, body);
}

/** Take the tunnel down. Idempotent; nothing answers the ticket until `enable`. */
export async function disableRemote(): Promise<RemoteStatus> {
  return post<RemoteStatus>(REMOTE_DISABLE_PATH, {});
}

/**
 * Reach another machine: bind a loopback port here that is its proxy. With
 * `pairing` as `<ticket>-<code>` the code is redeemed for that machine's key
 * and stored; omitted, the last ticket is dialled with the stored key.
 */
export async function connectRemote(body: Partial<RemoteConnectBody> = {}): Promise<RemoteConnectResponse> {
  return post<RemoteConnectResponse>(REMOTE_CONNECT_PATH, body);
}

/** Close the loopback port. Idempotent; the stored pairing stays. */
export async function disconnectRemote(): Promise<RemoteStatus> {
  return post<RemoteStatus>(REMOTE_DISCONNECT_PATH, {});
}

/**
 * Stop the far machine's daemon through the tunnel, then disconnect.
 *
 * A one-way door: nothing brings that daemon back except someone at the
 * machine. The daemon requires the confirmation word in the body, exactly as
 * the proxy route it forwards to does, so this cannot be reached by an
 * accidental empty POST — the caller has already asked the person.
 */
export async function killRemote(): Promise<RemoteStatus> {
  return post<RemoteStatus>(REMOTE_KILL_PATH, { confirm: 'shutdown' });
}

/**
 * Offer a code that hands one more device a key of its own.
 *
 * Nothing else about the session changes: the flags it was enabled with, the
 * ticket, and every device already using it are left exactly as they were.
 * Requires the tunnel to be up — a `409` naming the commands that fix it
 * otherwise — and answers with the same shape `enable` does, because what a
 * person is shown is the same thing.
 */
export async function inviteRemote(): Promise<RemoteEnableResponse> {
  return post<RemoteEnableResponse>(REMOTE_INVITE_PATH, {});
}

/**
 * Stop admitting one device and forget it, leaving every other device untouched.
 *
 * Works with the tunnel down, and must: a laptop is lost at a moment nobody
 * chose. Forgetting one this machine never held is a `200` with
 * `forgotten: false`, not a `404` — the outcome asked for is that no key is
 * held under that name, and none is.
 */
export async function forgetDevice(device: string): Promise<RemoteForgotten> {
  return del<RemoteForgotten>(`${REMOTE_DEVICES_PATH}/${encodeURIComponent(device)}`);
}
